full-moon/src/ast/luau.rs

```diff
@@ -183,6 +169,100 @@ pub enum TypeInfo {
     },
 }
 
+/// A union type, such as `string | number`.
+#[derive(Clone, Debug, Display, PartialEq, Node)]
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[display(fmt = "{}{types}", "display_option(leading)")]
+pub struct TypeUnion {
+    pub(crate) leading: Option<TokenReference>,
+    pub(crate) types: Punctuated<TypeInfo>,
+}
+
+impl TypeUnion {
+    /// Creates a new Union from the given types and optional leading pipe.
+    pub fn new(leading: Option<TokenReference>, types: Punctuated<TypeInfo>) -> Self {
+        Self { leading, types }
+    }
+
+    /// Returns a new Union with the given types.
+    pub fn with_types(self, types: Punctuated<TypeInfo>) -> Self {
+        Self { types, ..self }
+    }
+
+    /// Returns a new Union with the given leading pipe.
+    pub fn with_leading(self, leading: TokenReference) -> Self {
+        Self {
+            leading: Some(leading),
+            ..self
+        }
+    }
+
+    /// Returns a new Union without a leading pipe.
+    pub fn without_leading(self) -> Self {
+        Self {
+            leading: None,
+            ..self
+        }
+    }
```

Comment:
I think it'd be a simpler api surface and align more closely with rust conventions for `with_leading` to take an option instead of separating to 2 methods.